google/resource_composer_environment_test.go

```diff
@@ -0,0 +1,506 @@
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"google.golang.org/api/composer/v1"
+	"log"
+	"strings"
+)
+
+const environmentPrefix = "tf-composer-testenv"
+const networkPrefix = "tf-composer-testnet"
+
+func init() {
+	resource.AddTestSweepers("gcp_composer_environment", &resource.Sweeper{
+		Name: "gcp_composer_environment",
+		F:    testSweepComposerResources,
+	})
+}
+
+// Checks environment creation with minimum required information.
+func TestAccComposerEnvironment_basic(t *testing.T) {
+	t.Parallel()
+
+	envName := acctest.RandomWithPrefix(environmentPrefix)
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccComposerEnvironmentDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccComposerEnvironment_basic(envName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("google_composer_environment.test", "config.0.airflow_uri"),
+					resource.TestCheckResourceAttrSet("google_composer_environment.test", "config.0.gke_cluster"),
+					resource.TestCheckResourceAttrSet("google_composer_environment.test", "config.0.node_count"),
+					resource.TestCheckResourceAttrSet("google_composer_environment.test", "config.0.node_config.0.zone"),
+					resource.TestCheckResourceAttrSet("google_composer_environment.test", "config.0.node_config.0.machine_type")),
+			},
+			{
+				ResourceName:      "google_composer_environment.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				ResourceName:      "google_composer_environment.test",
+				ImportState:       true,
+				ImportStateId:     fmt.Sprintf("projects/%s/locations/%s/environments/%s", getTestProjectFromEnv(), "us-central1", envName),
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+// Checks that all updatable fields can be updated in one apply
+// (PATCH for Environments only is per-field)
+func TestAccComposerEnvironment_update(t *testing.T) {
+	t.Parallel()
+
+	envName := acctest.RandomWithPrefix(environmentPrefix)
+	var env composer.Environment
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccComposerEnvironmentDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccComposerEnvironment_basic(envName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComposerEnvironmentExists("google_composer_environment.test", &env),
+				),
+			},
+			{
+				Config: testAccComposerEnvironment_update(envName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComposerEnvironmentExists("google_composer_environment.test", &env),
+				),
+			},
+			{
+				ResourceName:      "google_composer_environment.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+
+	if env.Config == nil || env.Config.SoftwareConfig == nil {
```

Comment:
A lot of these checks are also unnecessary due to the import test (in general, anything that needs to be checked can almost always be covered by a combination of import tests and TestCheckResourceAttrs on computed fields)